### PR TITLE
Framework: Add action to prevent LFS files from being added

### PR DESCRIPTION
**Pull request by sebrowne**
@trilinos/framework 

## Motivation
Adding LFS pointers causes headaches with GitHub quotas and some downstream customers doing mirroring.

### CODE REVIEW

**Comment by sebrowne** [comment]
I realize that this has the same issue @bartlettroscoe mentioned about ctest vs. actions vs. hooks as here: https://github.com/trilinos/Trilinos/pull/12111#issuecomment-1692455329

**Comment by sebrowne** [comment]
I really haven't tested this to see if it works.  Will do so on this PR by adding (or at least trying to add) LFS pointers, then removing.  Though now that I think about it, we may need to check each commit....

**Comment by bartlettroscoe** [comment]
@sebrowne, were you able to test locally at least?

File: .github/workflows/detect-git-lfs.yml [hunk 1]
```diff
@@ -17,5 +17,5 @@ jobs:
       run: |
         git diff -U0 --ignore-all-space $(git merge-base origin/${{ github.event.pull_request.base.ref }} ${{ github.event.pull_request.head.sha }}) \
         | grep "^oid sha236" \
-        || ( echo "No pointers found!" ; exit 0 ) \
-        && ( echo "Git LFS pointers found; not allowed to add LFS files to Trilinos!" ; exit 1)
+        || { echo "No pointers found!" ; exit 0 } \
+        && { echo "Git LFS pointers found; not allowed to add LFS files to Trilinos!" ; exit 1
```
Comment (bartlettroscoe) [inline .github/workflows/detect-git-lfs.yml:21]: This should be a script that takes arguments.  Then this could be tested and run locally.

Comment (sebrowne) [reply .github/workflows/detect-git-lfs.yml]: I can do that.  Yes I did test the overall logic locally.  However, I think I want it to march through all of the intermediary commits and run this on each one.  Will make a script and get back.

**Comment by sebrowne** [comment]
Okay, test cases:

1.  No LFS pointers in any commits -- pass
2.  LFS pointer in HEAD -- fail
3.  LFS pointer in a commit, removed in HEAD -- fail

When taken together, this forces the committer of the LFS pointer to eradicate it from their history prior to the check passing.

**Comment by sebrowne** [comment]
If this looks satisfactory, please feel free to merge it.  I've tested it to my own satisfaction.

**Review by bartlettroscoe — CHANGES_REQUESTED** [review]
I think this can leave out commits from the analysis if the 'develop' branch has been merged into the topic 'branch' (which is common with some developers as a way to get up-to-date with 'develop').

File: commonTools/test/utilities/check-lfs-oids.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash    
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+claimed_start_of_branch=${1}
+end_of_branch=${2}
+start_of_branch=$(git merge-base ${claimed_start_of_branch} ${end_of_branch})
```
[the remark this answers is not in the document]
Comment (sebrowne) [reply commonTools/test/utilities/check-lfs-oids.sh]: Fixed, good call

File: commonTools/test/utilities/check-lfs-oids.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+for commit in $(git log --format=%H ${start_of_branch}..${end_of_branch})
```
[the remark this answers is not in the document]
Comment (sebrowne) [reply commonTools/test/utilities/check-lfs-oids.sh]: Fixed, good call.

File: commonTools/test/utilities/check-lfs-oids.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash    
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+claimed_start_of_branch=${1}
+end_of_branch=${2}
```
[the remark this answers is not in the document]
Comment (sebrowne) [reply commonTools/test/utilities/check-lfs-oids.sh]: I changed it to `source_branch`, how's that?

File: commonTools/test/utilities/check-lfs-oids.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash    
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+target_branch=${1}
+source_branch=${2}
+
+for commit in $(git log --format=%H ${source_branch} --not ${target_branch})
```
Comment (bartlettroscoe) [inline commonTools/test/utilities/check-lfs-oids.sh:10]: That is correct.  That should iterate over the all of the commits unique to the topic branch (source branch) not already in the target branch.

**Review by bartlettroscoe — APPROVED** [review]
Looks good to me.  Thanks for considering the suggestions.

**Comment by sebrowne** [comment]
@bartlettroscoe this works!  See #12203.  Thank you for the review!

**Pull request merged**